File: 107.BinaryTreeLevelOrderTraversalII/traverse.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ans;

        if (!root) {
            return ans;
        }

        int currLength = 0;
        int nextLength = 1;
        vector<int> level;

        queue<TreeNode*> q;
        q.push(root);

        vector<vector<int>> allLevel;
        while (!q.empty()) {
            if (currLength == 0) {
                allLevel.push_back(level);
                level.clear();
                currLength = nextLength;
                nextLength = 0;
            }

            TreeNode *node = q.front();
            q.pop();
            level.push_back(node->val);

            if (node->left) {
                q.push(node->left);
                nextLength++;
            }
            if (node->right) {
                q.push(node->right);
                nextLength++;
            }

            currLength--;
        }

        allLevel.push_back(level);

        for (int i = int(allLevel.size())-1; i >= 1; i--) {
            ans.push_back(allLevel[i]);
        }

        return ans;
    }
};

File: 107.BinaryTreeLevelOrderTraversalII/Traverse.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        if (root == null) {
            return new ArrayList<List<Integer>>();
        }

        int currLength = 0;
        int nextLength = 1;
        List<Integer> level = new ArrayList<Integer>();

        Deque<TreeNode> queue = new ArrayDeque<TreeNode>();
        queue.add(root);

        List<List<Integer>> allLevel = new ArrayList<List<Integer>>();
        while (!queue.isEmpty()) {
            if (currLength == 0) {
                allLevel.add(level);
                level = new ArrayList<Integer>();
                currLength = nextLength;
                nextLength = 0;
            }

            TreeNode node = queue.poll();
            level.add(node.val);
            currLength--;

            if (node.left != null) {
                queue.add(node.left);
                nextLength++;
            }
            if (node.right != null) {
                queue.add(node.right);
                nextLength++;
            }
        }

        allLevel.add(level);

        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        for (int i = allLevel.size() - 1; i >= 1; i--) {
            ans.add(allLevel.get(i));
        }

        return ans;
    }
}
